### PR TITLE
Implements new Warehouse design

### PR DESCRIPTION
**Pull request by aeslaughter**

### CODE REVIEW

**Comment by aeslaughter** [comment]
```
slauae@slaughter:~/projects/moose/framework (ic-wh-5680)$ git diff --shortstat upstream/devel -- .
 121 files changed, 3175 insertions(+), 3354 deletions(-)
```

**Comment by aeslaughter** [comment]
YAK patch is ready to go.

File: framework/include/base/BlockRestrictable.h [hunk 1]
```diff
@@ -175,7 +177,13 @@ class BlockRestrictable
    * Return all of the SubdomainIDs for the mesh
    * @return A set of all subdomians for the entire domain
    */
-  const std::set<SubdomainID> & meshBlockIDs();
+  const std::set<SubdomainID> & meshBlockIDs() const;
+
+  /**
+   * Returns true if this object has been restricted to a boundary
```
Comment (permcody) [inline framework/include/base/BlockRestrictable.h]: comment is wrong

File: framework/src/base/ComputeNodalAuxVarsThread.C [hunk 2]
```diff
@@ -51,13 +51,18 @@ ComputeNodalAuxVarsThread::onNode(ConstNodeRange::const_iterator & node_it)
...
+
+  // Loop over all SubdomainIDs for the curnent node, if an AuxKernel is active on this block then compute it.
   const std::set<SubdomainID> & block_ids = _sys.mesh().getNodeBlockIds(*node);
   for (std::set<SubdomainID>::const_iterator block_it = block_ids.begin(); block_it != block_ids.end(); ++block_it)
   {
-    for (std::vector<AuxKernel*>::const_iterator aux_it = _auxs[_tid].activeBlockNodalKernels(*block_it).begin();
-         aux_it != _auxs[_tid].activeBlockNodalKernels(*block_it).end();
-         ++aux_it)
-      (*aux_it)->compute();
+    std::map<SubdomainID, std::vector<MooseSharedPointer<AuxKernel> > >::const_iterator iter = block_kernels.find(*block_it);
```
Comment (permcody) [inline framework/src/base/ComputeNodalAuxVarsThread.C]: We really should consider using unordered maps inside of these compute threaded areas.

File: framework/src/base/FEProblem.C [hunk 3]
```diff
@@ -2292,7 +2259,7 @@ FEProblem::computeAuxiliaryKernels(const ExecFlagType & type)
 void
 FEProblem::computeUserObjects(const ExecFlagType & type, const UserObjectWarehouse::GROUP & group)
 {
-  // Perform Residual/Jacobian setups
+  // Perform Residual/Jacobain setups
```
Comment (permcody) [inline framework/src/base/FEProblem.C]: I like the correct spelling better.

File: framework/src/base/NonlinearSystem.C [hunk 4]
```diff
@@ -325,39 +322,38 @@ NonlinearSystem::stopSolve()
...
+    _kernels.initialSetup(tid);
+    _dirac_kernels.initialSetup(tid);
+    if (_doing_dg)
+      _dg_kernels.initialSetup(tid);
+
+    _dampers.initialSetup(tid);
+    _integrated_bcs.initialSetup(tid);
   }
+  _scalar_kernels.initialSetup();
+  _constraints.initialSetup();
```
Comment (permcody) [inline framework/src/base/NonlinearSystem.C]: Isn't constraints threaded? I see it being resized above?

File: framework/src/base/NonlinearSystem.C [hunk 5]
```diff
@@ -544,11 +540,16 @@ NonlinearSystem::addKernel(const std::string & kernel_name, const std::string &
...
-
     // Add the kernel to the warehouse
-    _kernels[tid].addKernel(kernel, blk_ids);
+    _kernels.addObject(kernel, tid);
+
+    // Store time/non-time kernels separately
+    MooseSharedPointer<TimeKernel> t_kernel = MooseSharedNamespace::dynamic_pointer_cast<TimeKernel>(kernel);
+    if (t_kernel)
+      _time_kernels.addObject(kernel, tid);
+    else
```
Comment (permcody) [inline framework/src/base/NonlinearSystem.C]: Wait, this doesn't seem right. How do we know that this kernel is either a time kernel or a nodal time kernel. What if it's neither?

File: framework/src/base/NonlinearSystem.C [hunk 6]
```diff
@@ -584,48 +585,49 @@ NonlinearSystem::addNodalKernel(const std::string & kernel_name, const std::stri
...
+    const std::set<BoundaryID> & boundary_ids = bc->boundaryIDs();
+    _vars[tid].addBoundaryVar(boundary_ids, &bc->variable());
+
+    // PresetNodalBC
     MooseSharedPointer<PresetNodalBC> pnbc = MooseSharedNamespace::dynamic_pointer_cast<PresetNodalBC>(bc);
-    if (pnbc.get())
-      _bcs[tid].addPresetNodalBC(boundary_ids, pnbc);
+    if (pnbc)
+      if (tid == 0)
+        _preset_nodal_bcs.addObject(pnbc);
```
Comment (permcody) [inline framework/src/base/NonlinearSystem.C]: So the old preset nodal bcs were threaded but the new ones aren't? why?

File: framework/src/bcs/BoundaryCondition.C [hunk 7]
```diff
@@ -29,6 +29,7 @@ InputParameters validParams<BoundaryCondition>()
   params.addParam<bool>("use_displaced_mesh", false, "Whether or not this object should use the displaced mesh for computation.  Note that in the case this is true but no displacements are provided in the Mesh block the undisplaced mesh will still be used.");
   params.addParamNamesToGroup("use_displaced_mesh", "Advanced");
 
+  params.declareControllable("enable");
```
Comment (permcody) [inline framework/src/bcs/BoundaryCondition.C]: Is this parameter inherited, or is it first declared here? If it's the later, how do you apply any of the other attributes like required, range-checked, etc?

File: framework/src/executioners/Transient.C [hunk 8]
```diff
@@ -660,7 +663,7 @@ Transient::preExecute()
   */
 
   // Add time period start times to sync times
-  const std::vector<TimePeriod *> time_periods = _problem.getTimePeriods();
+  const std::vector<TimePeriodOld *> time_periods = _problem.getTimePeriods();
```
Comment (permcody) [inline framework/src/executioners/Transient.C]: Is `TimePeriodOld` used anywhere? Should we just remove it?

File: framework/include/base/ExecuteMooseObjectStorage.h [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+template<typename T>
+void
+ExecuteMooseObjectStorage<T>::addObject(MooseSharedPointer<T> object, THREAD_ID tid/*=0*/)
+{
+  // Update list of all objects
+  MooseObjectStorage<T>::addObject(object, tid);
+
+  // Update the execute flag lists of objects
+  MooseSharedPointer<SetupInterface> ptr = MooseSharedNamespace::dynamic_pointer_cast<SetupInterface>(object);
+  if (ptr)
```
Comment (permcody) [inline framework/include/base/ExecuteMooseObjectStorage.h]: There should probably be an error when this fails. I don't think we want MOOSE to silently fail to add objects to containers.

File: framework/include/base/MooseObjectStorage.h [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+  checkThreadID(tid);
+  typename std::vector<MooseSharedPointer<T> >::const_iterator it;
+  for (it = _active_objects[tid].begin(); it != _active_objects[tid].end(); ++it)
+    (*it)->subdomainSetup();
+}
+
+
+template<typename T>
+void
+MooseObjectStorage<T>::jacobianSetup(THREAD_ID tid/* = 0*/) const
```
Comment (permcody) [inline framework/include/base/MooseObjectStorage.h]: I'm a little confused. This class which is the "threaded copy store" also has time-based setup methods `jacobianSetup`, `residualSetup`, etc. Aren't those types of objects typically associated with execution flags? I see those same methods defined in `ExecuteMooseObjectStorage`.

**Comment by permcody** [comment]
One of the current design flaws in MOOSE is that when we have objects that exist in multiple exec stores we call the same setup methods on them for each location they are stored. The typical case is when we have an object with flags like `"initial timestep_end"`. That object will have `initialSetup` called multiple times. I don't believe this design flaw was addressed in this new layout.

There is still a lot of code that has to mess around with global, block, and boundary restrictable containers throughout. This is less trivial to fix in the new design since we could potentially have thousands of subdomains. Still I wonder if we could do something to abstract out all the handling of the various types. Maybe not, just a thought.

@idaholab/moose-developers - I encourage you to take a look at this design. Even though I've performed the review this is a huge change worth looking at. @friedmud - it would be nice to get your thoughts as well.

File: framework/src/base/NonlinearSystem.C (hunk 4 above)
Comment (aeslaughter) [inline framework/src/base/NonlinearSystem.C]: Constraints are not threaded and it is not resized.

https://github.com/idaholab/moose/blob/devel/framework/src/base/NonlinearSystem.C#L879
https://github.com/idaholab/moose/blob/devel/framework/src/base/NonlinearSystem.C#L902

File: framework/src/base/NonlinearSystem.C (hunk 5 above)
Comment (aeslaughter) [inline framework/src/base/NonlinearSystem.C]: Nodal kernels are in a different warehouse that has not been updated yet.

File: framework/src/base/NonlinearSystem.C (hunk 6 above)
Comment (aeslaughter) [inline framework/src/base/NonlinearSystem.C]: There was threaded copies stored in the warehouse, they were not computed threaded.

https://github.com/idaholab/moose/blob/devel/framework/src/base/NonlinearSystem.C#L822

File: framework/src/bcs/BoundaryCondition.C (hunk 7 above)
Comment (aeslaughter) [inline framework/src/bcs/BoundaryCondition.C]: This parameter is defined in MooseObject.C, it is by default set to "true."

File: framework/src/executioners/Transient.C (hunk 8 above)
Comment (aeslaughter) [inline framework/src/executioners/Transient.C]: There is tests and I believe it is used by some applications. This PR will deprecate it and I will remove it after I get all the warehouses updated.

File: framework/include/base/ExecuteMooseObjectStorage.h (hunk 9 above)
Comment (aeslaughter) [inline framework/include/base/ExecuteMooseObjectStorage.h]: I will add a mooseError.

File: framework/include/base/MooseObjectStorage.h (hunk 10 above)
Comment (aeslaughter) [inline framework/include/base/MooseObjectStorage.h]: "execute_on" is not a one to one relationship with the setup methods. For example, Kernels have all the setup methods but are executed by Moose as needed. All objects have the setup methods, only some objects are executed with "execute_on" flags.

**Comment by aeslaughter** [comment]
@permcody I did address the multiple calls to the setup methods (this PR actually fixes a memory leak in TransientMultiApp). If you look at `MooseObjectStorage` and `ExecuteMooseObjectStorage` the calls to initialSetup happen on the vector of all objects, not the execute based storage.

**Comment by aeslaughter** [comment]
>There is still a lot of code that has to mess around with global, block, and boundary restrictable containers throughout. 

In general, all of the boundary/block restricted checks occur in MooseObjectStorageBase. However, the InitialConditionStorage is a special case, it does some additional checks for boundary/block restricted stuff, but this is strictly for error checking: to make sure that multiple ICs are not set on the same block/boundary etc.

**Comment by friedmud** [comment]
I plan on reviewing this tomorrow.  Finally finished the semester today... trying to catch up...

**Comment by permcody** [comment]
Oh almost forgot to write down my big comment. I don't think we should switch away from using "Warehouse" in our object names. It's one of those things that's been in MOOSE since the beginning and changing away from it arbitrary.

**Comment by aeslaughter** [comment]
@permcody I will switch the names to Warehouse, storage just seemed more natural to me for no apparent reason.

**Comment by permcody** [comment]
I'd be fine with that if this was a new project but we came up with the Factory/Warehouse pair in the original design of MOOSE years ago. You build objects in a factory and store them in a warehouse. Our users have used it for years so we are more or less stuck with the name when there isn't a driving reason to change it.

**Pull request merged**